### PR TITLE
API prefactor

### PR DESCRIPTION
**Pull request by snopoke**
@emord 

This moves some code around and then removes some old versions of API's that are not in use anymore.

The API usage for the last 2 months looks like this:

```
 401732 v0.4/form
 317416 v0.5/case
 258097 v0.5/form
  98605 v0.4/case
  55661 v0.5/user
  45426 v0.5/group
  12394 v0.3/location_internal
  12334 v0.4/fixture
   2873 v0.5/sso
   1444 v0.5/application
   1130 v0.4/user
    297 v0.5/data-forwarding
     40 v0.4/group
     36 v0.5/sms_user_registration
     33 v0.5/configurablereportdata
     28 v0.5/domain_forms
     24 v0.5/web-user
     21 v0.5/stock_transaction
     15 v0.5/simplereportconfiguration
     13 v0.3/case
     11 v0.5/fixture_internal
      6 v0.5/fixture
      5 v0.5/project_space_metadata
      4 v0.5/casehttps:
      4 v0.4/web-user
      3 v0.4/data-forwarding
      2 v0.6/web-user
      2 v0.5/attachment
      2 v0.4/cases
      1 v0.5/users
      1 v0.5/forms
      1 v0.5/cases
      1 v0.4/sms_user_registration
      1 v0.1/case
```

### CODE REVIEW

File: corehq/apps/api/resources/auth.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        PASSED_AUTH = 'is_authenticated'
+
+        def dummy(request, domain, **kwargs):
+            return PASSED_AUTH
+
+        wrapped_dummy = dummy
+        for wrapper in wrappers:
+            wrapped_dummy = wrapper(wrapped_dummy)
+
+        if not kwargs.has_key('domain'):
```
Comment (dimagimon) [inline corehq/apps/api/resources/auth.py]: W601 .has_key() is deprecated, use 'in'

File: corehq/apps/api/resources/meta.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+            return False
+
+        return super(HQThrottle, self).should_be_throttled(identifier, **kwargs)
+
+
+class CustomResourceMeta(object):
+    authorization = ReadOnlyAuthorization()
+    authentication = LoginAndDomainAuthentication()
+    serializer = CustomXMLSerializer()
+    default_format='application/json'
```
Comment (dimagimon) [inline corehq/apps/api/resources/meta.py]: E225 missing whitespace around operator

File: corehq/apps/api/resources/meta.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    throttle = HQThrottle(throttle_at=getattr(settings, 'CCHQ_API_THROTTLE_REQUESTS', 25),
+                             timeframe=getattr(settings, 'CCHQ_API_THROTTLE_TIMEFRAME', 15))
```
Comment (dimagimon) [inline corehq/apps/api/resources/meta.py]: E127 continuation line over-indented for visual indent

File: corehq/apps/api/urls.py [hunk 4]
```diff
@@ -11,35 +11,11 @@
...
-        v0_1.WebUserResource,
-        v0_2.CommCareCaseResource,
-        v0_1.XFormInstanceResource,
-        FixtureResource,
-        DomainMetadataResource,
-    )),
     ((0, 3), (
-        v0_1.CommCareUserResource,
-        v0_1.WebUserResource,
         v0_3.CommCareCaseResource,
```
Comment (snopoke) [inline corehq/apps/api/urls.py]: This is still used by 2 project (but only a few requests). We could try and communicate out to them that we're removing this API and they should use v0.4 or v0.5.

File: corehq/apps/api/urls.py [hunk 5]
```diff
@@ -11,35 +11,11 @@
...
-        v0_3.XFormInstanceResource,
-        FixtureResource,
-        DomainMetadataResource,
-        locations.v0_1.LocationResource,
-        locations.v0_1.InternalLocationResource,
         ProductResource,
```
Comment (snopoke) [inline corehq/apps/api/urls.py]: I don't see any usage for this but it's for LMIS so not sure if that's needed still: https://confluence.dimagi.com/display/lmis/Read+API

File: corehq/apps/api/urls.py (hunk 4 above)
Comment (emord) [inline corehq/apps/api/urls.py]: Should we block on merging until some comms is sent to them?

Comment (snopoke) [inline corehq/apps/api/urls.py]: Can always remove the in a later PR

Comment (emord) [inline corehq/apps/api/urls.py]: Oh. nevermind. I misread what line this comment was on

**Comment by emord** [comment]
build fail looks real:

```
======================================================================
ERROR: test suite for <class 'corehq.apps.api.tests.InternalLocationResourceTest'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 209, in run
    self.setUp()
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/vendor/lib/python2.7/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/mnt/commcare-hq-ro/corehq/apps/api/tests.py", line 105, in setUpClass
    cls.list_endpoint = cls._get_list_endpoint()
  File "/mnt/commcare-hq-ro/corehq/apps/api/tests.py", line 120, in _get_list_endpoint
    resource_name=cls.resource._meta.resource_name))
  File "/vendor/lib/python2.7/site-packages/django/core/urlresolvers.py", line 551, in reverse
    return iri_to_uri(resolver._reverse_with_prefix(view, prefix, *args, **kwargs))
  File "/vendor/lib/python2.7/site-packages/django/core/urlresolvers.py", line 468, in _reverse_with_prefix
    (lookup_view_s, args, kwargs, len(patterns), patterns))
NoReverseMatch: Reverse for 'api_dispatch_list' with arguments '()' ....
```

**Comment by snopoke** [comment]
:white_check_mark:

**Pull request merged**